File: scripts/apply_sec002_modular_exact.py
#!/usr/bin/env python3
"""Apply the SEC-002 ordered unified-diff hunks exactly across a modular clasp clone.

Safety properties:
- no fuzzy matching;
- every hunk preimage must match exactly once across SERVER_JS files;
- patch order comes from qa/sec002_private_delivery_bundle_manifest.json;
- only the expected core modules may change;
- English LAB / Memory Match modules are forbidden targets;
- emits a machine-readable report with before/after SHA-256 values.

This helper edits only the local clasp clone. It never calls clasp and never writes remote state.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

ALLOWED_CHANGED_BASENAMES = {
    "01_Router",
    "14_Notas_Cierre_Certificados",
    "20_Inscripcion_Ventas_Matricula",
    "21_Pagos_Banco_CONAPE",
}
FORBIDDEN_NAME_MARKERS = ("ENGLISH_LAB", "MEMORY_MATCH")
PRIVATE_ENDPOINTS = (
    "descargarMiCertificadoPrivado",
    "descargarDocumentoExtraPrivado",
    "descargarComprobantePagoPrivado",
    "descargarMatriculaFirmadaPrivada",
)


@dataclass
class Hunk:
    patch_path: str
    index: int
    header: str
    old_lines: list[str]
    new_lines: list[str]


def sha256_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def sha256_path(path: Path) -> str:
    return sha256_bytes(path.read_bytes())


def normalize_newlines(text: str) -> str:
    return text.replace("\r\n", "\n").replace("\r", "\n")


def detect_newline(raw: bytes) -> str:
    return "\r\n" if raw.count(b"\r\n") > raw.count(b"\n") // 2 else "\n"


def source_files(apps_dir: Path) -> list[Path]:
    files: list[Path] = []
    for path in apps_dir.rglob("*"):
        if not path.is_file():
            continue
        if path.suffix.lower() not in {".js", ".gs"}:
            continue
        files.append(path)
    return sorted(files)


def parse_patch(path: Path, display_path: str) -> list[Hunk]:
    lines = normalize_newlines(path.read_text(encoding="utf-8")).split("\n")
    hunks: list[Hunk] = []
    i = 0
    hidx = 0
    while i < len(lines):
        if not lines[i].startswith("@@ "):
            i += 1
            continue
        header = lines[i]
        i += 1
        old_lines: list[str] = []
        new_lines: list[str] = []
        while i < len(lines) and not lines[i].startswith("@@ "):
            line = lines[i]
            if line.startswith("--- ") or line.startswith("+++ "):
                break
            if line.startswith("\\ No newline at end of file"):
                i += 1
                continue
            if not line:
                # A truly empty unified-diff content line still has a prefix.
                # A bare empty line here is the split artifact at EOF/outside hunk.
                i += 1
                continue
            prefix, body = line[0], line[1:]
            if prefix == " ":
                old_lines.append(body)
                new_lines.append(body)
            elif prefix == "-":
                old_lines.append(body)
            elif prefix == "+":
                new_lines.append(body)
            else:
                break
            i += 1
        hidx += 1
        if not old_lines:
            raise RuntimeError(f"Hunk without preimage is not allowed: {display_path} {header}")
        hunks.append(Hunk(display_path, hidx, header, old_lines, new_lines))
    return hunks


def count_occurrences(text: str, needle: str) -> int:
    if not needle:
        return 0
    count = 0
    start = 0
    while True:
        pos = text.find(needle, start)
        if pos < 0:
            return count
        count += 1
        start = pos + 1


def useful_anchor(lines: Iterable[str]) -> str:
    candidates = [line for line in lines if line.strip() and len(line.strip()) >= 12]
    if not candidates:
        return ""
    candidates.sort(key=lambda s: (len(s.strip()), s), reverse=True)
    return candidates[0].strip()


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--apps-dir", required=True)
    ap.add_argument("--repo-root", required=True)
    ap.add_argument("--report", required=True)
    args = ap.parse_args()

    apps_dir = Path(args.apps_dir).resolve()
    repo_root = Path(args.repo_root).resolve()
    report_path = Path(args.report).resolve()
    manifest_path = repo_root / "qa" / "sec002_private_delivery_bundle_manifest.json"

    if not apps_dir.is_dir():
        raise RuntimeError(f"apps-dir does not exist: {apps_dir}")
    if not manifest_path.is_file():
        raise RuntimeError(f"manifest missing: {manifest_path}")

    manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    ordered = sorted(manifest["ordered_deltas"], key=lambda item: int(item["order"]))

    files = source_files(apps_dir)
    if not files:
        raise RuntimeError("No .js/.gs server files found in clasp clone")

    original_bytes = {p: p.read_bytes() for p in files}
    texts = {p: normalize_newlines(original_bytes[p].decode("utf-8-sig")) for p in files}
    newline_by_file = {p: detect_newline(original_bytes[p]) for p in files}

    total_expected_hunks = sum(int(item["expected_hunks"]) for item in ordered)
    applied: list[dict] = []
    failures: list[dict] = []

    for item in ordered:
        rel_patch = str(item["path"])
        patch_path = repo_root / rel_patch
        hunks = parse_patch(patch_path, rel_patch)
        if len(hunks) != int(item["expected_hunks"]):
            raise RuntimeError(
                f"Hunk contract mismatch for {rel_patch}: expected={item['expected_hunks']} parsed={len(hunks)}"
            )

        for hunk in hunks:
            old_text = "\n".join(hunk.old_lines)
            new_text = "\n".join(hunk.new_lines)
            matches: list[tuple[Path, int]] = []
            for path, text in texts.items():
                occurrences = count_occurrences(text, old_text)
                if occurrences:
                    matches.append((path, occurrences))

            exact_total = sum(n for _, n in matches)
            if exact_total != 1:
                anchor = useful_anchor(hunk.old_lines)
                anchor_hits = []
                if anchor:
                    for path, text in texts.items():
                        n = count_occurrences(text, anchor)
                        if n:
                            anchor_hits.append({"file": str(path.relative_to(apps_dir)), "count": n})
                failures.append(
                    {
                        "patch": hunk.patch_path,
                        "hunk": hunk.index,
                        "header": hunk.header,
                        "exact_match_count": exact_total,
                        "exact_match_files": [
                            {"file": str(p.relative_to(apps_dir)), "count": n} for p, n in matches
                        ],
                        "anchor": anchor,
                        "anchor_hits": anchor_hits,
                    }
                )
                break

            target = next(p for p, n in matches if n == 1)
            base = target.stem
            upper_name = target.name.upper()
            if base not in ALLOWED_CHANGED_BASENAMES:
                failures.append(
                    {
                        "patch": hunk.patch_path,
                        "hunk": hunk.index,
                        "header": hunk.header,
                        "error": "unexpected_target_module",
                        "file": str(target.relative_to(apps_dir)),
                        "allowed": sorted(ALLOWED_CHANGED_BASENAMES),
                    }
                )
                break
            if any(marker in upper_name for marker in FORBIDDEN_NAME_MARKERS):
                failures.append(
                    {
                        "patch": hunk.patch_path,
                        "hunk": hunk.index,
                        "header": hunk.header,
                        "error": "forbidden_lab_memory_target",
                        "file": str(target.relative_to(apps_dir)),
                    }
                )
                break

            texts[target] = texts[target].replace(old_text, new_text, 1)
            applied.append(
                {
                    "patch": hunk.patch_path,
                    "hunk": hunk.index,
                    "header": hunk.header,
                    "file": str(target.relative_to(apps_dir)),
                }
            )
        if failures:
            break

    changed_files: list[Path] = []
    before_sha: dict[str, str] = {}
    after_sha: dict[str, str] = {}

    if not failures and len(applied) != total_expected_hunks:
        failures.append(
            {
                "error": "applied_hunk_total_mismatch",
                "expected": total_expected_hunks,
                "observed": len(applied),
            }
        )

    if not failures:
        for path in files:
            original_text = normalize_newlines(original_bytes[path].decode("utf-8-sig"))
            if texts[path] == original_text:
                continue
            changed_files.append(path)

        changed_bases = {p.stem for p in changed_files}
        if not changed_bases.issubset(ALLOWED_CHANGED_BASENAMES):
            failures.append(
                {
                    "error": "changed_file_allowlist_violation",
                    "observed": sorted(changed_bases),
                    "allowed": sorted(ALLOWED_CHANGED_BASENAMES),
                }
            )

    if not failures:
        combined = "\n".join(texts[p] for p in files)
        endpoint_counts = {
            endpoint: len(re.findall(rf"function\s+{re.escape(endpoint)}\s*\(", combined))
            for endpoint in PRIVATE_ENDPOINTS
        }
        bad_endpoint_counts = {k: v for k, v in endpoint_counts.items() if v != 1}
        if bad_endpoint_counts:
            failures.append(
                {
                    "error": "private_endpoint_definition_count",
                    "counts": endpoint_counts,
                }
            )
    else:
        endpoint_counts = {}

    if not failures:
        for path in changed_files:
            rel = str(path.relative_to(apps_dir)).replace("\\", "/")
            before_sha[rel] = sha256_bytes(original_bytes[path])
            newline = newline_by_file[path]
            out = texts[path].replace("\n", newline)
            path.write_text(out, encoding="utf-8", newline="")
            after_sha[rel] = sha256_path(path)

    report = {
        "ok": not failures,
        "mode": "SEC002_MODULAR_EXACT_REBASE_V1",
        "apps_dir": str(apps_dir),
        "manifest": str(manifest_path),
        "total_expected_hunks": total_expected_hunks,
        "applied_hunks": len(applied),
        "applied": applied,
        "changed_files": [str(p.relative_to(apps_dir)).replace("\\", "/") for p in changed_files],
        "before_sha256": before_sha,
        "after_sha256": after_sha,
        "endpoint_definition_counts": endpoint_counts,
        "failures": failures,
    }
    report_path.parent.mkdir(parents=True, exist_ok=True)
    report_path.write_text(json.dumps(report, indent=2, ensure_ascii=False), encoding="utf-8")

    if failures:
        print("SEC002_MODULAR_REBASE_STOP")
        print(json.dumps(failures[0], ensure_ascii=False))
        print(f"REPORT={report_path}")
        return 2

    print("SEC002_MODULAR_REBASE_PASS")
    print(f"APPLIED_HUNKS={len(applied)}")
    print("CHANGED_FILES=" + ",".join(report["changed_files"]))
    for rel in report["changed_files"]:
        print(f"SHA {rel} before={before_sha[rel]} after={after_sha[rel]}")
    print(f"REPORT={report_path}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"SEC002_MODULAR_REBASE_FATAL: {exc}", file=sys.stderr)
        raise

File: scripts/apply_sec002_modular_exact.mjs
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';

const ALLOWED_CHANGED_BASENAMES = new Set([
  '01_Router',
  '14_Notas_Cierre_Certificados',
  '20_Inscripcion_Ventas_Matricula',
  '21_Pagos_Banco_CONAPE',
]);
const FORBIDDEN_NAME_MARKERS = ['ENGLISH_LAB', 'MEMORY_MATCH'];
const PRIVATE_ENDPOINTS = [
  'descargarMiCertificadoPrivado',
  'descargarDocumentoExtraPrivado',
  'descargarComprobantePagoPrivado',
  'descargarMatriculaFirmadaPrivada',
];

function argValue(name) {
  const i = process.argv.indexOf(name);
  if (i < 0 || i + 1 >= process.argv.length) throw new Error(`Missing ${name}`);
  return process.argv[i + 1];
}
function normalize(text) { return text.replace(/\r\n/g, '\n').replace(/\r/g, '\n'); }
function sha256(buf) { return crypto.createHash('sha256').update(buf).digest('hex'); }
function countOccurrences(text, needle) {
  if (!needle) return 0;
  let count = 0, start = 0;
  while (true) {
    const p = text.indexOf(needle, start);
    if (p < 0) return count;
    count++; start = p + 1;
  }
}
function walk(dir) {
  const out = [];
  for (const ent of fs.readdirSync(dir, {withFileTypes:true})) {
    const p = path.join(dir, ent.name);
    if (ent.isDirectory()) out.push(...walk(p));
    else if (ent.isFile() && /\.(?:js|gs)$/i.test(ent.name)) out.push(p);
  }
  return out.sort();
}
function parsePatch(filePath, displayPath) {
  const lines = normalize(fs.readFileSync(filePath, 'utf8')).split('\n');
  const hunks = [];
  let i = 0, hidx = 0;
  while (i < lines.length) {
    if (!lines[i].startsWith('@@ ')) { i++; continue; }
    const header = lines[i++];
    const oldLines = [], newLines = [];
    while (i < lines.length && !lines[i].startsWith('@@ ')) {
      const line = lines[i];
      if (line.startsWith('--- ') || line.startsWith('+++ ')) break;
      if (line.startsWith('\\ No newline at end of file')) { i++; continue; }
      if (!line) { i++; continue; }
      const prefix = line[0], body = line.slice(1);
      if (prefix === ' ') { oldLines.push(body); newLines.push(body); }
      else if (prefix === '-') oldLines.push(body);
      else if (prefix === '+') newLines.push(body);
      else break;
      i++;
    }
    hidx++;
    if (!oldLines.length) throw new Error(`Hunk without preimage: ${displayPath} ${header}`);
    hunks.push({patch:displayPath, hunk:hidx, header, oldLines, newLines});
  }
  return hunks;
}
function anchor(lines) {
  return [...lines].filter(x => x.trim().length >= 12).sort((a,b)=>b.trim().length-a.trim().length)[0]?.trim() || '';
}
function baseNoExt(file) { return path.basename(file).replace(/\.(?:js|gs)$/i, ''); }

const appsDir = path.resolve(argValue('--apps-dir'));
const repoRoot = path.resolve(argValue('--repo-root'));
const reportPath = path.resolve(argValue('--report'));
const manifestPath = path.join(repoRoot, 'qa', 'sec002_private_delivery_bundle_manifest.json');

const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
const ordered = [...manifest.ordered_deltas].sort((a,b)=>Number(a.order)-Number(b.order));
const files = walk(appsDir);
if (!files.length) throw new Error('No .js/.gs server files found in clasp clone');

const original = new Map();
const text = new Map();
const newline = new Map();
for (const f of files) {
  const b = fs.readFileSync(f);
  original.set(f, b);
  const s = b.toString('utf8').replace(/^\uFEFF/, '');
  text.set(f, normalize(s));
  newline.set(f, (b.toString('binary').match(/\r\n/g)||[]).length > (b.toString('binary').match(/(?<!\r)\n/g)||[]).length ? '\r\n' : '\n');
}

const totalExpected = ordered.reduce((n,x)=>n+Number(x.expected_hunks),0);
const applied = [];
const failures = [];

outer:
for (const item of ordered) {
  const patchRel = String(item.path);
  const hunks = parsePatch(path.join(repoRoot, patchRel), patchRel);
  if (hunks.length !== Number(item.expected_hunks)) throw new Error(`Hunk contract mismatch ${patchRel}: expected=${item.expected_hunks} parsed=${hunks.length}`);
  for (const h of hunks) {
    const oldText = h.oldLines.join('\n');
    const newText = h.newLines.join('\n');
    const matches = [];
    for (const f of files) {
      const n = countOccurrences(text.get(f), oldText);
      if (n) matches.push({file:f,count:n});
    }
    const exactTotal = matches.reduce((n,m)=>n+m.count,0);
    if (exactTotal !== 1) {
      const a = anchor(h.oldLines);
      const anchorHits = [];
      if (a) for (const f of files) { const n=countOccurrences(text.get(f),a); if(n) anchorHits.push({file:path.relative(appsDir,f).replaceAll('\\','/'),count:n}); }
      failures.push({patch:h.patch,hunk:h.hunk,header:h.header,exact_match_count:exactTotal,exact_match_files:matches.map(m=>({file:path.relative(appsDir,m.file).replaceAll('\\','/'),count:m.count})),anchor:a,anchor_hits:anchorHits});
      break outer;
    }
    const target = matches.find(m=>m.count===1).file;
    const base = baseNoExt(target);
    const upper = path.basename(target).toUpperCase();
    if (!ALLOWED_CHANGED_BASENAMES.has(base)) {
      failures.push({patch:h.patch,hunk:h.hunk,header:h.header,error:'unexpected_target_module',file:path.relative(appsDir,target).replaceAll('\\','/'),allowed:[...ALLOWED_CHANGED_BASENAMES].sort()});
      break outer;
    }
    if (FORBIDDEN_NAME_MARKERS.some(x=>upper.includes(x))) {
      failures.push({patch:h.patch,hunk:h.hunk,header:h.header,error:'forbidden_lab_memory_target',file:path.relative(appsDir,target).replaceAll('\\','/')});
      break outer;
    }
    text.set(target, text.get(target).replace(oldText,newText));
    applied.push({patch:h.patch,hunk:h.hunk,header:h.header,file:path.relative(appsDir,target).replaceAll('\\','/')});
  }
}

if (!failures.length && applied.length !== totalExpected) failures.push({error:'applied_hunk_total_mismatch',expected:totalExpected,observed:applied.length});

const changed = [];
if (!failures.length) {
  for (const f of files) {
    const orig = normalize(original.get(f).toString('utf8').replace(/^\uFEFF/,''));
    if (text.get(f) !== orig) changed.push(f);
  }
  const bad = changed.map(baseNoExt).filter(x=>!ALLOWED_CHANGED_BASENAMES.has(x));
  if (bad.length) failures.push({error:'changed_file_allowlist_violation',observed:bad,allowed:[...ALLOWED_CHANGED_BASENAMES].sort()});
}

let endpointCounts = {};
if (!failures.length) {
  const combined = files.map(f=>text.get(f)).join('\n');
  for (const endpoint of PRIVATE_ENDPOINTS) {
    const re = new RegExp(`function\\s+${endpoint}\\s*\\(`,'g');
    endpointCounts[endpoint] = (combined.match(re)||[]).length;
  }
  if (Object.values(endpointCounts).some(n=>n!==1)) failures.push({error:'private_endpoint_definition_count',counts:endpointCounts});
}

const beforeSha = {}, afterSha = {};
if (!failures.length) {
  for (const f of changed) {
    const rel = path.relative(appsDir,f).replaceAll('\\','/');
    beforeSha[rel] = sha256(original.get(f));
    const out = text.get(f).replace(/\n/g,newline.get(f));
    fs.writeFileSync(f,out,'utf8');
    afterSha[rel] = sha256(fs.readFileSync(f));
  }
}

const report = {
  ok: !failures.length,
  mode: 'SEC002_MODULAR_EXACT_REBASE_V1',
  apps_dir: appsDir,
  manifest: manifestPath,
  total_expected_hunks: totalExpected,
  applied_hunks: applied.length,
  applied,
  changed_files: changed.map(f=>path.relative(appsDir,f).replaceAll('\\','/')),
  before_sha256: beforeSha,
  after_sha256: afterSha,
  endpoint_definition_counts: endpointCounts,
  failures,
};
fs.mkdirSync(path.dirname(reportPath),{recursive:true});
fs.writeFileSync(reportPath,JSON.stringify(report,null,2),'utf8');

if (failures.length) {
  console.log('SEC002_MODULAR_REBASE_STOP');
  console.log(JSON.stringify(failures[0]));
  console.log(`REPORT=${reportPath}`);
  process.exit(2);
}
console.log('SEC002_MODULAR_REBASE_PASS');
console.log(`APPLIED_HUNKS=${applied.length}`);
console.log(`CHANGED_FILES=${report.changed_files.join(',')}`);
for (const rel of report.changed_files) console.log(`SHA ${rel} before=${beforeSha[rel]} after=${afterSha[rel]}`);
console.log(`REPORT=${reportPath}`);
